stylus-verifier/stylus-verifier-logic/src/docker.rs

```diff
@@ -74,65 +109,193 @@ async fn create_image(
+                    "error while building an image"
+                );
+                anyhow::bail!(
+                    "error while building an image; image_name={image_name}, output={output}"
+                );
+            }
+            Err(err) => {
+                let output = output.join("");
+                tracing::error!(
+                    image_name = image_name,
+                    output = output,
+                    error = err.to_string(),
+                    "unknown error while building an image"
+                );
+                anyhow::bail!("unknown error while building an image; image_name={image_name}, output={output}, error={err}");
+            }
+        }
+    }
 
     Ok(())
 }
 
-async fn run_in_docker_container(
-    cargo_stylus_version: &Version,
-    toolchain: &Version,
-    dir: &Path,
+async fn create_container(
+    docker: &Docker,
+    image_name: &str,
     command_line: &[&str],
 ) -> Result<String, anyhow::Error> {
-    let name = version_to_image_name(cargo_stylus_version, toolchain);
-    if !image_exists(&name).await? {
-        anyhow::bail!("Docker image {name} doesn't exist");
-    }
+    let container_suffix = Uuid::new_v4();
+    let container_name = format!(
+        "{}-{container_suffix}",
+        image_name.replace(|c: char| !c.is_alphanumeric(), "_")
+    );
+
+    let options = CreateContainerOptions {
+        name: container_name.clone(),
+        ..Default::default()
+    };
+    let config = container::Config {
+        image: Some(image_name),
+        working_dir: Some(WORKDIR),
+        host_config: Some(HostConfig {
+            network_mode: Some("host".to_string()),
+            auto_remove: Some(true),
+            ..Default::default()
+        }),
+        cmd: Some(command_line.into()),
+        ..Default::default()
+    };
 
-    let output = Command::new("docker")
-        .arg("run")
-        .arg("--network")
-        .arg("host")
-        .arg("-w")
-        .arg("/source")
-        .arg("-v")
-        .arg(format!("{}:/source", dir.as_os_str().to_str().unwrap()))
-        .arg("--rm")
-        .arg(name)
-        .args(command_line)
-        .output()
+    let container = docker.create_container(Some(options), config).await?;
+
+    Ok(container.id)
+}
+
+async fn copy_directory_to_container(
+    docker: &Docker,
+    container_id: &str,
+    dir: &Path,
+) -> Result<(), anyhow::Error> {
+    let tar = build_tar_from_directory(dir).context("building tar archive from directory")?;
+
+    let options = UploadToContainerOptions {
+        path: WORKDIR,
+        no_overwrite_dir_non_dir: "",
+    };
+    docker
+        .upload_to_container(container_id, Some(options), tar.into())
         .await
-        .context("failed to execute Docker run command")?;
+        .context("uploading tar archive to container")?;
+
+    Ok(())
+}
+
+async fn start_container(docker: &Docker, container_id: &str) -> Result<String, anyhow::Error> {
```

Comment:
nit:
it seems to not only start but run the container till the end (until output stream ends). Maybe `run_container`?

Reply:
Updated in [c5ce76d](https://github.com/blockscout/blockscout-rs/pull/1074/commits/c5ce76da327956834ca5426ac3c8dae0d13686ef)